### PR TITLE
Upload pasted image as JPEG if needed

### PR DESCRIPTION
**Pull request by davidmz**
Upload pasted image as JPEG if:
1) The PNG size is more than 50 KiB and
2) JPEG size is less than half the PNG size.

This algorithm keeps ever big drawing- and text screenshots as PNG and converts photos to JPEG.

### CODE REVIEW

File: src/components/create-post.jsx [hunk 1]
```diff
@@ -224,3 +224,51 @@ export default class CreatePost extends React.Component {
...
+  const src = window.URL.createObjectURL(blob);
+  try {
+    const image = await new Promise((resolve, reject) => {
+      const image = new Image();
+      image.onload = () => resolve(image);
+      image.onerror = () => reject(new Error('Cannot load image'));
+      image.src = src;
+    });
+
+    console.log(image.width,image.height); // eslint-disable-line
```
Comment (indeyets) [inline src/components/create-post.jsx]: Do we need this?

File: src/components/create-post.jsx [hunk 2]
```diff
@@ -224,3 +224,51 @@ export default class CreatePost extends React.Component {
...
+
+    const canvas = document.createElement('canvas');
+    canvas.width = image.width;
+    canvas.height = image.height;
+    canvas.getContext('2d').drawImage(image, 0, 0);
+
+    const jpeg = await new Promise((resolve) => canvas.toBlob(resolve, "image/jpeg", 0.9));
+    jpeg.name = blob.name.replace(/\.png$/, '.jpg');
+
+    console.log(jpeg.name, jpeg.size); // eslint-disable-line
```
Comment (indeyets) [inline src/components/create-post.jsx]: this one is left from initial debugging probably?

File: src/components/create-post.jsx [hunk 3]
```diff
@@ -224,3 +224,51 @@ export default class CreatePost extends React.Component {
...
+ * 2) JPEG size is less than half the PNG size.
+ *
+ * The returning promise is never failed and returns
+ * either JPEG or the original PNG.
+ *
+ * @param {Blob} blob
+ * @returns {Promise<Blob>}
+ */
+export async function makeJpegIfNeeded(blob) {
+  if (blob.type !== 'image/png' || blob.size < 100 * 1024) {
```
Comment (indeyets) [inline src/components/create-post.jsx]: why 100KiB limit? difference can still be pretty dramatic for smaller images

**Review by indeyets — CHANGES_REQUESTED** [review]

File: src/components/create-post.jsx (hunk 2 above)
Comment (davidmz) [reply src/components/create-post.jsx]: Of couse there are debug artifacts:) Thank you for check.

File: src/components/create-post.jsx (hunk 3 above)
Comment (davidmz) [reply src/components/create-post.jsx]: I thought that <100 KiB PNG won't hurt anyone, even if it is not optimal. We can lower this limit but i think it should be more than zero to keep some small but colored images (icons, drawed userpics) as PNG.

Comment (clbn) [reply src/components/create-post.jsx]: > <100 KiB PNG won't hurt anyone

Is it "PNG" though? I mean, is blob size equal PNG size?

File: src/components/create-post.jsx [hunk 4]
```diff
@@ -224,3 +224,47 @@ export default class CreatePost extends React.Component {
...
+
+    const canvas = document.createElement('canvas');
+    canvas.width = image.width;
+    canvas.height = image.height;
+    canvas.getContext('2d').drawImage(image, 0, 0);
+
+    const jpeg = await new Promise((resolve) => canvas.toBlob(resolve, "image/jpeg", 0.9));
+    jpeg.name = blob.name.replace(/\.png$/, '.jpg');
+
+    if (jpeg.size < blob.size / 2) {
```
Comment (clbn) [inline src/components/create-post.jsx:264]: Same question about `blob.size` — how big is it in comparison to PNG format? Isn't it much bigger? Or does browser encode it into PNG for us, so size is for PNG?

File: src/components/create-post.jsx (hunk 3 above)
Comment (davidmz) [reply src/components/create-post.jsx]: Yes, the pasted Blob _is_ a PNG-formatted image file made by browser (or OS? I dont know for sure). It is the same PNG that will be uploaded to our server.

File: src/components/create-post.jsx (hunk 4 above)
Comment (davidmz) [reply src/components/create-post.jsx:264]: As I say in other answer, yes, it is a PNG file size.

**Pull request merged**